undefined reference to symbol `''` (repro included)

### Remember to search before filing a new report

- [X] I searched for this bug and did not find it in the issue tracker, and furthermore, the title I used above will make this new bug report turn up in the search results for my query.

### Zig Version

0.9.0-dev.902+5eb4f3459

### Steps to Reproduce

Repo where you can repro: https://github.com/Jarred-Sumner/zig-ld-bug

1. Reference an `extern fn` from [s2n-tls](https://github.com/aws/s2n-tls) on macOS x64
2. `zig build-exe` with the `.a` files
3. This happens:
```
error(link): undefined reference to symbol ''
error(link):   first referenced in '/Users/jarred/Build/zig-ld-bug/zig-cache/o/c52829028565f17f9f891944f4f5646c/main.o'
error(link): undefined reference to symbol ''
error(link):   first referenced in '/Users/jarred/Build/zig-ld-bug/zig-cache/o/c52829028565f17f9f891944f4f5646c/main.o'
error(link): undefined reference to symbol ''
error(link):   first referenced in '/Users/jarred/Build/zig-ld-bug/zig-cache/o/c52829028565f17f9f891944f4f5646c/main.o'
error: UndefinedSymbolReference
```

### Expected Behavior

`zig build-exe` runs successfully and the executable executes successfully

`clang` is able to link the object files successfully.

### Actual Behavior

```
error(link): undefined reference to symbol ''
error(link):   first referenced in '/Users/jarred/Build/zig-ld-bug/zig-cache/o/c52829028565f17f9f891944f4f5646c/main.o'
error(link): undefined reference to symbol ''
error(link):   first referenced in '/Users/jarred/Build/zig-ld-bug/zig-cache/o/c52829028565f17f9f891944f4f5646c/main.o'
error(link): undefined reference to symbol ''
error(link):   first referenced in '/Users/jarred/Build/zig-ld-bug/zig-cache/o/c52829028565f17f9f891944f4f5646c/main.o'
error: UndefinedSymbolReference
```

If there are multiple `.o` files, it sometimes will report a different `.o` file in the error message (i think it's always the first one?)

--- a/src/link/MachO.zig
+++ b/src/link/MachO.zig
@@ -2311,7 +2311,7 @@ fn createDsoHandleAtom(self: *MachO) !void {
         nlist.n_desc = macho.N_WEAK_DEF;
         try self.globals.append(self.base.allocator, nlist);
 
-        _ = self.unresolved.fetchSwapRemove(resolv.where_index);
+        assert(self.unresolved.swapRemove(resolv.where_index));
 
         undef.* = .{
             .n_strx = 0,
@@ -2409,7 +2409,7 @@ fn resolveSymbolsInObject(self: *MachO, object_id: u16) !void {
                     const global = &self.globals.items[resolv.where_index];
 
                     if (symbolIsTentative(global.*)) {
-                        _ = self.tentatives.fetchSwapRemove(resolv.where_index);
+                        assert(self.tentatives.swapRemove(resolv.where_index));
                     } else if (!(symbolIsWeakDef(sym) or symbolIsPext(sym)) and
                         !(symbolIsWeakDef(global.*) or symbolIsPext(global.*)))
                     {
@@ -2437,7 +2437,7 @@ fn resolveSymbolsInObject(self: *MachO, object_id: u16) !void {
                         .n_desc = 0,
                         .n_value = 0,
                     };
-                    _ = self.unresolved.fetchSwapRemove(resolv.where_index);
+                    assert(self.unresolved.swapRemove(resolv.where_index));
                 },
             }
 
@@ -2496,6 +2496,8 @@ fn resolveSymbolsInObject(self: *MachO, object_id: u16) !void {
                         .n_value = sym.n_value,
                     });
                     _ = try self.tentatives.getOrPut(self.base.allocator, global_sym_index);
+                    assert(self.unresolved.swapRemove(resolv.where_index));
+
                     resolv.* = .{
                         .where = .global,
                         .where_index = global_sym_index,
@@ -2508,7 +2510,6 @@ fn resolveSymbolsInObject(self: *MachO, object_id: u16) !void {
                         .n_desc = 0,
                         .n_value = 0,
                     };
-                    _ = self.unresolved.fetchSwapRemove(resolv.where_index);
                 },
             }
         } else {
@@ -3412,7 +3413,7 @@ pub fn updateDeclExports(
                     const sym = &self.globals.items[resolv.where_index];
 
                     if (symbolIsTentative(sym.*)) {
-                        _ = self.tentatives.fetchSwapRemove(resolv.where_index);
+                        assert(self.tentatives.swapRemove(resolv.where_index));
                     } else if (!is_weak and !(symbolIsWeakDef(sym.*) or symbolIsPext(sym.*))) {
                         _ = try module.failed_exports.put(
                             module.gpa,
@@ -3438,7 +3439,7 @@ pub fn updateDeclExports(
                     continue;
                 },
                 .undef => {
-                    _ = self.unresolved.fetchSwapRemove(resolv.where_index);
+                    assert(self.unresolved.swapRemove(resolv.where_index));
                     _ = self.symbol_resolver.remove(n_strx);
                 },
             }

--- a/test/standalone/link_common_symbols/b.c
+++ b/test/standalone/link_common_symbols/b.c
@@ -1,5 +1,6 @@
 long i;
 int j = 2;
+int k;
 
 void incr_i() {
   i++;

--- a/test/standalone/link_common_symbols/build.zig
+++ b/test/standalone/link_common_symbols/build.zig
@@ -4,7 +4,7 @@ pub fn build(b: *Builder) void {
     const mode = b.standardReleaseOptions();
 
     const lib_a = b.addStaticLibrary("a", null);
-    lib_a.addCSourceFiles(&.{ "a.c", "b.c" }, &.{"-fcommon"});
+    lib_a.addCSourceFiles(&.{ "c.c", "a.c", "b.c" }, &.{"-fcommon"});
     lib_a.setBuildMode(mode);
 
     const test_exe = b.addTest("main.zig");

--- a/test/standalone/link_common_symbols/c.c
+++ b/test/standalone/link_common_symbols/c.c
@@ -1,0 +1,5 @@
+extern int k;
+
+int common_defined_externally() {
+  return k;
+}

--- a/test/standalone/link_common_symbols/main.zig
+++ b/test/standalone/link_common_symbols/main.zig
@@ -1,8 +1,13 @@
 const std = @import("std");
 const expect = std.testing.expect;
 
+extern fn common_defined_externally() c_int;
 extern fn incr_i() void;
 extern fn add_to_i_and_j(x: c_int) c_int;
+
+test "undef shadows common symbol: issue #9937" {
+    try expect(common_defined_externally() == 0);
+}
 
 test "import C common symbols" {
     incr_i();

